source/server/listener_manager_impl.cc

```diff
@@ -438,42 +438,25 @@ bool ListenerManagerImpl::addOrUpdateListenerInternal(
 
   ListenerImpl& new_listener_ref = *new_listener;
 
-  // We mandate that a listener with the same name must have the same configured address. This
-  // avoids confusion during updates and allows us to use the same bound address. Note that in
-  // the case of port 0 binding, the new listener will implicitly use the same bound port from
-  // the existing listener.
-  bool active_listener_exists = false;
-  bool warming_listener_exists = false;
-  if (existing_warming_listener != warming_listeners_.end() &&
-      *(*existing_warming_listener)->address() != *new_listener->address()) {
-    warming_listener_exists = true;
-  }
-  if (existing_active_listener != active_listeners_.end() &&
-      *(*existing_active_listener)->address() != *new_listener->address()) {
-    active_listener_exists = true;
-  }
-  if (active_listener_exists || warming_listener_exists) {
-    const std::string message =
-        fmt::format("error updating listener: '{}' has a different address '{}' from existing "
-                    "listener address '{}'",
-                    name, new_listener->address()->asString(),
-                    warming_listener_exists ? (*existing_warming_listener)->address()->asString()
-                                            : (*existing_active_listener)->address()->asString());
-    ENVOY_LOG(warn, "{}", message);
-    throw EnvoyException(message);
-  }
-
   bool added = false;
   if (existing_warming_listener != warming_listeners_.end()) {
     // In this case we can just replace inline.
     ASSERT(workers_started_);
     new_listener->debugLog("update warming listener");
-    new_listener->setSocketFactory((*existing_warming_listener)->getSocketFactory());
+    if (*(*existing_warming_listener)->address() != *new_listener->address()) {
+      setNewOrDrainingSocketFactory(name, config.address(), *new_listener, config.reuse_port());
```

Comment:
Can you add a test case to cover this branch?
https://storage.googleapis.com/envoy-pr/a07e91a/coverage/source/server/listener_manager_impl.cc.gcov.html